clang/lib/Basic/Targets/OSTargets.h

```diff
@@ -384,6 +384,9 @@ class LLVM_LIBRARY_VISIBILITY LinuxTargetInfo : public OSTargetInfo<Target> {
     } else {
         Builder.defineMacro("__gnu_linux__");
     }
+
+    if (Triple.isMusl())
+      Builder.defineMacro("__musl__", "1");
```

Comment:
Worth pointing out the Musl people seem to think this is fundamentally misguided: https://wiki.musl-libc.org/faq. Beyond that `__MUSL__` looks more similar to the existing `__GLIBC__`.

So probably not suitable for upstream, but for a one-off Apple branch where no-one is really using Musl other than you (hopefully) go ahead.

Reply:
Yes, I'm aware that they don't want to define a macro themselves. Unfortunately without having Clang define one here, there's no way for Clang's built-in headers to react to being used with musl, which is necessary if they want the system headers to take precedence in some cases.

To elaborate slightly on this issue this solves, `libc++` overrides some C library headers in order to make them C++ compatible — this is normal behaviour for C++ libraries, not some funky thing that's specific to `libc++` — and tries to `#include_next` the underlying C library header. With the rearranged include order that was checked in to Clang, that explodes because it ends up `#include_next`ing the Clang built-in header, which tries to `#include_next` the system header, but because it's at the end of the include order, that fails and we get a build failure.

The right solution is to do what happens on _every other_ platform where we want to prefer the system header, namely to detect the platform somehow and `#include_next` the system header from the Clang built-in header, instead of trying to mess with the include order and make it different just for one platform which is what was happening before. And that means we need to know we're building for a musl triple.

Reply:
Oh, as for `__MUSL__` vs `__musl__`, I didn't want to define the former _mainly_ because I think _that_ one should be defined by Musl (if they ever want to); as you say, it's more similar to the `__GLIBC__` macro (which contains the major version number for Glibc, and I think you'd want `__MUSL__` to do the same for Musl).

This Clang one has no way to tell which version of Musl is being used; it's really just about the triple.